### PR TITLE
Add `ABI0` support to Go `crypto/tls` eBPF tracer

### PR DESCRIPTION
**Pull request by mertyildiran**
Adds `vmlinux.h` for x86 and ARM64 which is generated by the command below in the machines with the corresponding architecture:

```
bpftool btf dump file /sys/kernel/btf/vmlinux format c > vmlinux_x86.h
bpftool btf dump file /sys/kernel/btf/vmlinux format c > vmlinux_arm64.h
```

that contains the declaration of `struct task_struct`. The struct is needed to get the Goroutine ID from thread-local storage. It's described in the `ABI0` specs.

Go `ABI0` (<=1.16) specification
https://go.dev/doc/asm

## `task_struct`, CO-RE and `cilium/ebpf`

~~The heap-like area for `struct task_struct` requires ring buffer. It means that; the PR increases the minimum supported Linux kernel version from `4.10` to `5.8`.~~

We read `task_struct` using [`BPF_CORE_READ`](https://nakryiko.com/posts/bpf-core-reference-guide/#bpf-core-read-1). Its member `thread` is a `thread_struct` and `thread_struct` is different in kernel versions older than or equal to `4.6`.

But `cilium/ebpf` does not support `.kconfig`:

```c
extern int LINUX_KERNEL_VERSION __kconfig;
```

Therefore; for now, we build object files per kernel version and load them accordingly.

### Possible TODOs for this subject;

 - `cilium/cilium` tags images per kernel version. So, do the same thing.
 - Wait `cilium/ebpf` to add `.kconfig` support. (uncertain)
 - Don't use `cilium/ebpf` or fork it.

### Tested

- [x] OpenSSL works on x86_64
- [x] OpenSSL works on ARM64
- [x] Go `ABI0` works on x86_64
- [x] Go `ABI0` works on ARM64
- [x] Go `ABIInternal` works on x86_64
- [x] Go `ABIInternal` works on ARM64

### Build Strategies

- [x] BUILDARCH=amd64 TARGETARCH=amd64
- [x] BUILDARCH=arm64v8 TARGETARCH=arm64v8
- [x] BUILDARCH=amd64 TARGETARCH=arm64v8
- [x] BUILDARCH=arm64v8 TARGETARCH=amd64

### Test Reference

https://github.com/up9inc/sock-shop-mizu-demo/pull/23

Screenshot from 2022-06-30 21-34-21

Screenshot from 2022-07-01 00-26-59

### CODE REVIEW

File: tap/tlstapper/bpf/go_uprobes.c [hunk 1]
```diff
@@ -55,10 +56,64 @@ Capstone Engine: https://www.capstone-engine.org/
...
+        return 0;
+    }
+
+    // We need another heap for task_struct in here.
+    // So we reserve from BPF_MAP_TYPE_RINGBUF defined above.
+    // It's different than the heap for tls_cunk in maps.h
+    // Using BPF_MAP_TYPE_PERCPU_ARRAY for task_struct makes the generator fail.
+    struct task_struct *task;
+    task = bpf_ringbuf_reserve(&task_struct_heap, sizeof(struct task_struct), 0);
+    if (!task) {
```
[the remark this answers is not in the document]
Comment (mertyildiran) [reply tap/tlstapper/bpf/go_uprobes.c]: @nimrod-up9 fixed with https://github.com/up9inc/mizu/commit/ca7cae20f7aeb7608f3afb3edcf0b80e87fe7bb3

File: tap/tlstapper/bpf/go_uprobes.c [hunk 2]
```diff
@@ -55,10 +56,64 @@ Capstone Engine: https://www.capstone-engine.org/
...
+        return 0;
+    }
+
+    // Get the task that currently assigned to this thread.
+    __u64 task_ptr = bpf_get_current_task();
+    if (!task_ptr) {
```
[the remark this answers is not in the document]
Comment (mertyildiran) [reply tap/tlstapper/bpf/go_uprobes.c]: @nimrod-up9 fixed with https://github.com/up9inc/mizu/commit/ca7cae20f7aeb7608f3afb3edcf0b80e87fe7bb3

File: tap/tlstapper/go_hooks.go [hunk 3]
```diff
@@ -75,6 +87,17 @@ func (s *goHooks) installHooks(bpfObjects *tlsTapperObjects, ex *link.Executable
 		s.goReadExProbes = append(s.goReadExProbes, probe)
 	}
 
+	// Pass goid and g struct offsets to an eBPF map to retrieve it in eBPF context
+	if err := bpfObjects.tlsTapperMaps.GoidOffsetsMap.Put(
```
[the remark this answers is not in the document]
Comment (mertyildiran) [reply tap/tlstapper/go_hooks.go]: @nimrod-up9 fixed with https://github.com/up9inc/mizu/commit/881f7e4a97356a62058f3da0176c4d8996df0bd6

**Comment by mertyildiran** [comment]
PR description is updated.

**Review by nimrod-up9 — COMMENTED** [review]
Approved, except maybe remove support for kernel 4.6. See slack discussion.

**Review by nimrod-up9 — DISMISSED** [review]

**Pull request merged**